service/resource/resourcegroup/resource.go

```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	currentResourceGroup, err := toResourceGroup(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredResourceGroup, err := toResourceGroup(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "checking if the resource group has to be deleted")
+
+	var resourceGroupToDelete *azureresource.Group
+	if currentResourceGroup != nil {
+		resourceGroupToDelete = desiredResourceGroup
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "checked if the resource group has to be deleted")
+
+	return resourceGroupToDelete, nil
+}
+
+// GetUpdateState returns nil for the create, delete and update states because
+// resource groups are not updated.
+func (r *Resource) GetUpdateState(obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
+	return nil, nil, nil, nil
+}
+
+// Name returns the resource name.
+func (r *Resource) Name() string {
+	return Name
+}
+
+// ProcessCreateState creates the resource group via the Azure API.
+func (r *Resource) ProcessCreateState(obj, createState interface{}) error {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	resourceGroupToCreate, err := toResourceGroup(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if resourceGroupToCreate != nil {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating the resource group in the Azure API")
+
+		groupClient, err := r.getGroupsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		_, err = groupClient.CreateOrUpdate(*resourceGroupToCreate.Name, *resourceGroupToCreate)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "created the resource group in the Azure API")
+
+	} else {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "the resource group already exists in the Azure API")
+	}
+
+	return nil
+}
+
+// ProcessDeleteState deletes the resource group via the Azure API.
+func (r *Resource) ProcessDeleteState(obj, deleteState interface{}) error {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	resourceGroupToDelete, err := toResourceGroup(deleteState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if resourceGroupToDelete != nil {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "deleting the resource group in the Azure API")
+
+		groupsClient, err := r.getGroupsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Delete the resource group which also deletes all resources it
+		// contains. We wait for the error channel while the deletion happens.
+		_, errchan := groupsClient.Delete(*resourceGroupToDelete.Name, nil)
+		err = <-errchan
+		if err != nil {
```

Comment:
Here we wait for the errchan to make sure there are no errors deleting the resource group. Deleting the group also deletes all resources it contains.

We should be able to use this to cleanly delete clusters without having orphan resources.

Reply:
Can we be sure this will never block forever?

Reply:
Uncertain, there should be a timeout on the API call. I'll double check that closes the error channel.